plugin/src/main/java/org/opensearch/ml/action/mcpserver/TransportMcpToolsRegisterOnNodesAction.java

```diff
@@ -0,0 +1,150 @@
+
+import com.google.common.collect.ImmutableMap;
+
+import io.modelcontextprotocol.server.McpServerFeatures;
+import io.modelcontextprotocol.spec.McpSchema;
+import lombok.extern.log4j.Log4j2;
+import reactor.core.publisher.Flux;
+import reactor.core.publisher.Mono;
+
+/**
+ * This class is responsible for registering tools on nodes.
+ */
+@Log4j2
+public class TransportMcpToolsRegisterOnNodesAction extends
+    TransportNodesAction<MLMcpToolsRegisterNodesRequest, MLMcpRegisterNodesResponse, MLMcpToolsRegisterNodeRequest, MLMcpRegisterNodeResponse> {
+    TransportService transportService;
+    ClusterService clusterService;
+    ThreadPool threadPool;
+    Client client;
+    NamedXContentRegistry xContentRegistry;
+    ToolFactoryWrapper toolFactoryWrapper;
+
+    private static final String INPUT_SCHEMA = "input_schema";
+
+    @Inject
+    public TransportMcpToolsRegisterOnNodesAction(
+        TransportService transportService,
+        ActionFilters actionFilters,
+        ClusterService clusterService,
+        ThreadPool threadPool,
+        Client client,
+        NamedXContentRegistry xContentRegistry,
+        ToolFactoryWrapper toolFactoryWrapper
+    ) {
+        super(
+            MLMcpToolsRegisterOnNodesAction.NAME,
+            threadPool,
+            clusterService,
+            transportService,
+            actionFilters,
+            MLMcpToolsRegisterNodesRequest::new,
+            MLMcpToolsRegisterNodeRequest::new,
+            ThreadPool.Names.MANAGEMENT,
+            MLMcpRegisterNodeResponse.class
+        );
+        this.transportService = transportService;
+        this.clusterService = clusterService;
+        this.threadPool = threadPool;
+        this.client = client;
+        this.xContentRegistry = xContentRegistry;
+        this.toolFactoryWrapper = toolFactoryWrapper;
+    }
+
+    @Override
+    protected MLMcpRegisterNodesResponse newResponse(
+        MLMcpToolsRegisterNodesRequest nodesRequest,
+        List<MLMcpRegisterNodeResponse> responses,
+        List<FailedNodeException> failures
+    ) {
+        return new MLMcpRegisterNodesResponse(clusterService.getClusterName(), responses, failures);
+    }
+
+    @Override
+    protected MLMcpToolsRegisterNodeRequest newNodeRequest(MLMcpToolsRegisterNodesRequest request) {
+        return new MLMcpToolsRegisterNodeRequest(request.getMcpTools());
+    }
+
+    @Override
+    protected MLMcpRegisterNodeResponse newNodeResponse(StreamInput in) throws IOException {
+        return new MLMcpRegisterNodeResponse(in);
+    }
+
+    @Override
+    protected MLMcpRegisterNodeResponse nodeOperation(MLMcpToolsRegisterNodeRequest request) {
+        return registerToolsOnNode(request.getMcpTools());
+    }
+
+    private MLMcpRegisterNodeResponse registerToolsOnNode(McpTools mcpTools) {
+        Flux.fromStream(mcpTools.getTools().stream()).flatMap(tool -> {
+            // check if user request contains tools that not in our system.
```

Comment:
What if the tool not in our system ? Should we handle this edge case ?

Reply:
This is already validated in the rest action: https://github.com/opensearch-project/ml-commons/blob/main/plugin/src/main/java/org/opensearch/ml/rest/mcpserver/RestMLRegisterMcpToolsAction.java#L89